pkg/controller/infinispan/infinispan_controller.go

```diff
@@ -203,6 +203,26 @@ func (r *ReconcileInfinispan) Reconcile(request reconcile.Request) (reconcile.Re
 		}
 	}
 
+	// Persistent volume claim for custom variables
+	if infinispan.Spec.Container.CustomLibrariesPVName != "" {
+		customPvc := &corev1.PersistentVolumeClaim{}
+		err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: infinispan.Namespace, Name: infinispan.Name + "-custom-libraries"}, customPvc)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				reqLogger.Info(fmt.Sprintf("Creating '%s'", infinispan.Spec.Container.CustomLibrariesPVName))
+				customPvc.Name = infinispan.Name + "-custom-libraries"
+				customPvc.Namespace = infinispan.Namespace
+				customPvc.Spec.AccessModes = []corev1.PersistentVolumeAccessMode{"ReadOnlyMany"}
+				customPvc.Spec.Resources.Requests = corev1.ResourceList{corev1.ResourceStorage: consts.DefaultPVSize}
+				customPvc.Spec.VolumeName = infinispan.Spec.Container.CustomLibrariesPVName
+				err = r.client.Create(context.TODO(), customPvc)
```

Comment:
Not sure that we need to create PVC if it not present

Reply:
I guess we need it, this pvc is for multiple read access and it's one per cluster so it cannot be created by the statefulset with the pvc template mechanism.

Reply:
Yes, but I'm wonder how customer will upload libraries to this PV and how server will handle this data
If it's already provided PV with uploaded libraries and linked to the SS it will work, but not sure that created PV by operator should work and handle by server without rebooting

Reply:
Yeah the PV must be provided by the user.
Operator doesn't create PV, this is up to the user: it should be pretty straightforward to do that with NFS, also GKE provides some tools for that, or he can rely on an ad-hoc data importer image.
Once the PV is prepared user just have to configure the PV name into the infinispan RC.

Reply:
Yes, thanks. It would be nice to place all storage logic inside SS reconciler.

Reply:
> Operator doesn't create PV, this is up to the user: it should be pretty straightforward to do that with NFS, also GKE provides some tools for that, or he can rely on an ad-hoc data importer image.

@rigazilla Do you have a link to any docs or blogs on how todo this?

Reply:
I'm more familiar with NFS, steps are:
- setup an nfs folder reachable by the ispn nodes
- represent your nfs folder in k8s, ie:
```
apiVersion: v1
kind: PersistentVolume
metadata:
  name: example-custom-libraries
spec:
  storageClass: "custom-libraries"
  accessModes:
  - ReadOnlyMany
  capacity:
    storage: 1Gi
  nfs:
    path: /opt/nfs/custom-libraries
    server: mynode.mydomain.com
  persistentVolumeReclaimPolicy: Retain
  volumeMode: Filesystem
```
- specify the PV name in the infinispan CR

For GKE I've found [this](https://stackoverflow.com/questions/57754103/how-to-pre-populate-a-readonlymany-persistent-volume), though I didn't dig into too much

Creating an ad hoc pod should be as easy as run a busybox image with the PV mounted and scp there the custom jars. Seems there's also a [general solution](https://github.com/kubevirt/containerized-data-importer).

Reply:
Should the user be providing a reference to `PersistentVolumClaim` instead of a `PersistentVolume`? The latter seems too low level.

There is a one-to-one mapping between a PVC and a PV, so if we create the PVC for the user this means that the read-only contents can only be used by a single Infinispan CR at once.

Consider a case where a company that has multiple Infinispan clusters, but they want to use the same scripts/marshalling/db-drivers in multiple clusters. If we create a PVC based upon the provided volume, it will be necessary for a volume to be created per cluster. However, if we allow the user to provide a `ReadOnlyMany` PVC we can mount that on all of the clusters.

Reply:
Good point @ryanemerson. 
I would like to keep the creation of the pvc on the operator side though, how about this policy:
- user provides the `volumeName`
- operator looks for the pvc with name `volumeName-infinispan-custom-libraries` and:
      - create it if doesn't exist 
      - otherwise reuses the existent
could it work?

Reply:
> I would like to keep the creation of the pvc on the operator side though, how about this policy:

Any particular reason? Creating a PVC is a higher level abstraction than a PV, so IMO this is more user friendly.

What's the advantage of allowing a user to provide a PV instead of a PVC?

> * operator looks for the pvc with name `volumeName-infinispan-custom-libraries` and:
>   - create it if doesn't exist
>   - otherwise reuses the existent
>   could it work?

Imo this seems harder to explain to a user then simply provide the name of an existing PVC.

Reply:
> > I would like to keep the creation of the pvc on the operator side though, how about this policy:
> 
> Any particular reason? Creating a PVC is a higher level abstraction than a PV, so IMO this is more user friendly.
> 
> What's the advantage of allowing a user to provide a PV instead of a PVC?

Even using the PVC the user must provide the PV name in the PVC, this because he wants to mount a specific volume and isn't asking for generic storage. I'm just suggesting to let the user to only specify the volume name and let the operator do the rest of the PVC configuration.

> 
> > * operator looks for the pvc with name `volumeName-infinispan-custom-libraries` and:
> >   
> >   * create it if doesn't exist
> >   * otherwise reuses the existent
> >     could it work?
> 
> IMO this is harder to explain to a user then "provide the name of an existing PVC".

he also needs to create the PVC if it doesn't exist and it's something that we could easily do for him since we know how we want to use it.

so basically the instruction for the user should be "provide the name of the PV with your jars"

Reply:
> Even using the PVC the user must provide the PV name in the PVC, this because he wants to mount a specific volume and isn't asking for generic storage.

That doesn't have to be the case. With a PVC the user can utilise generic storage as per the k8 cluster defaults.

For example, consider the busybox way of populating the data that you described. The user doesn't care what type of storage is used, so they just create a PVC with `ReadWriteOnce`, the PVC and volume are then mounted to a pod where they copy the required the files, the pod is then shutdown. This PVC is then mounted by the operator as readOnly in one or more Infinispan clusters.

Side note, but this is something that we could simplify with the CLI.

> he also needs to create the PVC if it doesn't exist and it's something that we could easily do for him since we know how we want to use it.

It's possible that the user just creates a PVC and names that, no need to create a volume.

I'm not sure, but in some environments doesn't the creation of a Volume require more permissions than a PVC?

> so basically the instruction for the user should be "provide the name of the PV with your jars"

Fair point. I was thinking that the user would need to be aware of this behaviour, but the operator can just do this behind the scenes.

Reply:
@ryanemerson I though it over, your approach is ok, let go ahead and leave the pvc creation to the user for now.

Reply:
Thanks for changing :+1:

This should greatly simplify creating a test for deployments, as we can do the following using the default volume type:

1. Create PVC
2. Mount PVC to pod
3. Copy files to pod
4. Shutdown pod
5. Start server with pvc and verify that the provided task etc is available via REST.